packages/gatsby/src/redux/reducers/redirects.js

```diff
@@ -1,11 +1,35 @@
 const _ = require(`lodash`)
 
+const createRedirectWithSlash = redirect => {
+  const { fromPath, toPath } = redirect
+  const isFile = pathPart => /\..+$/.test(pathPart)
+  const addSlashRedirect = (...parts) => parts.map(part => part.split(`/`).pop()).some(part => !isFile(part))
+
+  if (addSlashRedirect(fromPath, toPath)) {
+    return {
+      ...redirect,
+      ...(isFile(fromPath) ? {} : {
+        fromPath: `${fromPath}/`,
+      }),
+      ...(isFile(toPath) ? {} : {
+        toPath: `${toPath}/`,
+      }),
+    }
+  }
+
+  return []
+}
+
+/*
+ * Explicitly add redirect with and without trailing slash
+ */
 module.exports = (state = [], action) => {
   switch (action.type) {
     case `CREATE_REDIRECT`: {
-      if (!state.some(redirect => _.isEqual(redirect, action.payload))) {
-        // Add redirect only if it wasn't yet added to prevent duplicates
-        return [...state, action.payload]
+      const payload = [action.payload].concat(createRedirectWithSlash(action.payload))
```

Comment:
Does it make sense to add the logic here at the reducer level? I thought this made the most sense, but 👀 appreciated.